### DISCUSSION TITLE
Generated types for a view

### DISCUSSION THREAD

**Discussion by cdimitroulas** [supabase/supabase#14151] [Questions]
Hi there, wasn't sure if this was a bug or potential improvement or if this is a misunderstanding of mine about postgres views.

When a table has non-nullable columns, the generated Typescript types correctly reflect this. For example:
```ts
export interface Database {
  public: {
    Tables: {
      profiles: {
        Row: {
          avatarurl: string | null;
          bio: string | null;
          id: string; // non-nullable field
          joinedat: string; // non-nullable field
          username: string | null;
          website: string | null;
        };
        // etc...
```

However, when I created a view in the public schema I noticed that ALL columns were marked as nullable even though the underlying base table's columns are non-nullable. This was my migration file:
```sql
create schema if not exists internal;
grant usage on schema internal to postgres, service_role, dashboard_user;

create table internal.books (
  id uuid not null default gen_random_uui(),
  title text unique NOT NULL,
  author text NOT NULL,
  total_sales int NOT NULL,
  unique(title)
);

create view public.popular_books as
    select *
    from internal.books
    where total_sales > 1000000;
```

And this is the resulting Typescript type (all the the columns should not be nullable, but they are ALL marked as nullable):
```ts
    Views: {
      popular_books: {
        Row: {
          author: string | null;
          id: string | null;
          title: string | null;
          total_sales: number | null;
        };
```

Is this the expected result or should the type correctly reflect the underlying `books` schema which has all columns as non-nullable?

**Comment by ggounot** [comment]
I just ran into the same issue. It [seems like](https://dba.stackexchange.com/questions/285302/why-columns-in-views-are-always-nullable) it's a Postgres limitation.

**Comment by indymaat** [comment]
Refer to the [Helper Types documentation](https://supabase.com/docs/reference/javascript/typescript-support#helper-types) for guidance on overriding the `supabase/database-generated.types.ts` file generated by Supabase to tailor typings to your needs, including addressing the issue of nullable fields in views.

@cdimitroulas your code could look something like:
```
// supabase/database.types.ts
import { PostgrestError } from "@supabase/supabase-js";
import { MergeDeep } from "type-fest";
import { Database as DatabaseGenerated } from "./database-generated.types";

export type Database = MergeDeep<
  DatabaseGenerated,
  {
    public: {
      Views: {
        popular_books: {
          Row: {
            // Force the TypeScript compiler to think they are non-nullable fields
            author: string;
            id: string;
            title: string;
            total_sales: number;
          };
        };
      };
    };
  }
>;

export type TableRows<T extends keyof Database["public"]["Tables"]> =
  Database["public"]["Tables"][T]["Row"];
export type TableInserts<T extends keyof Database["public"]["Tables"]> =
  Database["public"]["Tables"][T]["Insert"];

export type Views<T extends keyof Database["public"]["Views"]> =
  Database["public"]["Views"][T]["Row"];

export type Enums<T extends keyof Database["public"]["Enums"]> =
  Database["public"]["Enums"][T];

export type DbResult<T> = T extends PromiseLike<infer U> ? U : never;
export type DbResultOk<T> = T extends PromiseLike<{ data: infer U }>
  ? Exclude<U, null>
  : never;
export type DbResultErr = PostgrestError;

export type DbResult<T> = T extends PromiseLike<infer U> ? U : never;
export type DbResultOk<T> = T extends PromiseLike<{ data: infer U }>
  ? Exclude<U, null>
  : never;
export type DbResultErr = PostgrestError;
```

They also let you access the actual typings easier. For example you could then do:
```
let popularBook: Views<"popular_books"> = // ...
// Instead of:
let popularBook: Database['public']['Views']['popular_books']['Row'] = // ...
```

**Reply by ianpaschal** [reply]
Any idea why the type generation offers a `Tables<'table_name'>` helper but no `Views<'table_name'>` helper?

**Comment by Xananax** [comment]
This blog post gives a good way of extending the types (not mine, I'm unrelated to the author): https://dev.to/nasheomirro/handling-view-types-from-supabase-with-typescript-41hj

**Comment by jacob-8** [comment]
I like the idea of using MergeDeep, but no matter how I try to implement (using v4.10.3 at the moment), I always end up with the result being `type Database = never` and can't use it all. I get this result regardless of project, using it exactly as demonstrated in the docs. Does anyone else have this issue?

**Reply by Xananax** [reply]
You need to set `strictNullChecks` to `true` in `tsconfig.json`:
```json
{
  "compilerOptions": {
    "target": "ES2022",
   ...
    "strictNullChecks": true,
   ...
```
2 weeks too late but hopefully it helps someone

**Reply by jacob-8** [reply]
@Xananax thank you very much! This is a wonderful help and should be mentioned in the documentation. Sadly, I won't be able to implement right away as using strictNullChecks induces 500+ errors. But it's good to know how it all works.

**Reply by jacob-8** [reply]
I submitted a PR to update the docs.

**Reply by jacob-8** [reply]
Ok, I tried using `strictNullChecks` and got fed up after a day of having to do this `let foo: string | null = null`. I like some of the theory of `strictNullChecks`, but it's an absolute disaster in any pre-existing app. So I wrote a script to take my generated types file, remove the Json type, read an `augments.types.ts` file and update the type of matching field between the two. Then it adds the imports from `augments.types.ts` and exports to a 3rd file I use for my types. Here's the script: https://gist.github.com/jacob-8/bde0e56629ba62a38a4c44a3feac851e

**Comment by Lukasvo** [comment]
I ran into the same issue with view types being overly nullable. Here’s a TypeScript utility that merges the view’s fields with the original table’s nullability, so you get the computed fields (like total_paid) but keep the correct nullability for the base columns:

```
type MergeWithOriginalNullability<
  Original,
  View
> = {
  [K in keyof Original]: Original[K];
} & {
    [K in Exclude<keyof View, keyof Original>]: View[K];
  };

type ReservationWithTotalPaid = MergeWithOriginalNullability<
  Database['public']['Tables']['reservations']['Row'],
  Database['public']['Views']['reservations_with_total_paid']['Row']
>;
```

This way, you get the best of both worlds: correct nullability for your base columns, and all the computed fields from your view.

Hope this helps! 🚀

**Comment by olivier-deschenes** [comment]
This type helper solved the problem for me.

```ts
type PreserveNullability<T> = null extends T ? T : Exclude<T, null>;

type TransformToView<Table, View> = {
  [K in Extract<keyof View, keyof Table>]: PreserveNullability<Table[K]>;
};
```

Used like:

```ts
type TransformedView = TransformToView<
  Tables<"YourTableName">,
  Tables<"YourTableViewName">
>
```

Example:

```ts
type UserTable = {
  id: number;              // Non-nullable (unique to table, so excluded)
  username: string;        // Non-nullable
  email: string | null;    // Nullable
  age: number | null;      // Nullable
  bio: string;             // Non-nullable (but will be excluded if not in view)
};

type UserView = {
  username: string | null; // Nullable in view
  email: string | null;    // Nullable in view
  age: number | null;      // Nullable in view
  // Note: 'id' and 'bio' are intentionally missing, so they'll be excluded
};

type UserTransformedView = TransformToView<UserTable, UserView>;
```

`UserTransformedView` will be of type:

```ts
type UserTransformedView = {
  username: string;        // Non-nullable (from table)
  email: string | null;    // Nullable (from table)
  age: number | null;      // Nullable (from table)
};
```

**Comment by cwharris** [comment]
I threw this together today.

I'm mapping table columns on to view columns and preservable nullability. This is only being used for intellisense, so I did _not_ vet this carefully. Maybe someone will find it useful as a starting point.

> [!caution]
> There's a lot that could go wrong here, and I haven't tested it thoroughly.

<img width="895" height="363" alt="image" src="https://github.com/user-attachments/assets/1df39d5a-033c-412d-a256-573b2c2079c6" />

### Database Transforms
```ts
import type { Database as SupaDatabase } from '$lib/supabase.types.ts';
import type { DatabaseWithTransforms, TransformNullability } from '$lib/database.helpers';

export type Database = DatabaseWithTransforms<SupaDatabase, [
  TransformNullability<SupaDatabase['app']['Views']['user_capabilities']['Row'], [
    SupaDatabase['app']['Tables']['org_role_capabilities']['Row'],
    SupaDatabase['app']['Tables']['org_role_bindings']['Row']
  ]>
]>;
```
### The Helper Types
```ts
type PreserveNullability<T> = null extends T ? T : Exclude<T, null>;

type TransformFromSources<
  Sources extends readonly any[],
  View
> = Sources extends readonly [infer First, ...infer Rest]
  ? {
      [K in keyof View]: K extends keyof First
        ? PreserveNullability<First[K]>
        : TransformFromSources<Rest, View>[K];
    }
  : View;

// Enhanced Override type that auto-detects source arrays
type TransformNullability<TOriginal, TNew> = {
  original: TOriginal;
  replacement: TNew extends readonly any[] 
    ? TransformFromSources<TNew, TOriginal>
    : TNew;
};

type ApplyOverrides<TBase, TOverrides extends readonly TransformNullability<any, any>[]> = 
  TOverrides extends readonly [infer First, ...infer Rest]
    ? First extends TransformNullability<infer TOriginal, any>
      ? Rest extends readonly TransformNullability<any, any>[]
        ? ApplyOverrides<
            ReplaceDeep<TBase, TOriginal, First['replacement']>,
            Rest
          >
        : ReplaceDeep<TBase, TOriginal, First['replacement']>
      : TBase
    : TBase;

type ReplaceDeep<TBase, TOriginal, TNew> = {
  [K in keyof TBase]: TBase[K] extends TOriginal
    ? TNew
    : TBase[K] extends object
    ? ReplaceDeep<TBase[K], TOriginal, TNew>
    : TBase[K];
};

type DatabaseWithTransforms<TBase, TOverrides extends readonly TransformNullability<any, any>[]> = 
  ApplyOverrides<TBase, TOverrides>;
```
